gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/BulkResultIterator.java

```diff
@@ -0,0 +1,115 @@
+  private List<String> header;
+  private int columnSize;
+  private int lineCount = 0; // this is different than currentFileRowCount. cvs file has header
+  private List<String> preLoadedLine = null;
+
+  public BulkResultIterator(BulkConnection conn, FileIdVO fileIdVO, int retryLimit) {
+    log.info("create BulkResultIterator: " + fileIdVO);
+    this.conn = conn;
+    this.fileIdVO = fileIdVO;
+    this.retryLimit = retryLimit;
+  }
+
+  /**
+   * read first data record from cvsReader and initiate header
+   * not supposed to do it in constructor function, for delay creating file stream
+   */
+  private void initHeader() {
+    this.header = this.nextLine(); // first line is header
+    this.columnSize = this.header.size();
+    this.preLoadedLine = this.nextLine(); // initialize: buffer one record data
+  }
+
+  private List<String> nextLine() {
+    Exception exception = null;
+    for (int i = 0; i < retryLimit; i++) {
+      try {
+        if (this.csvReader == null) {
+          this.csvReader = openAndSeekCsvReader();
+        }
+        List<String> line = this.csvReader.nextRecord();
+        this.lineCount++;
+        return line;
+      } catch (InputStreamCSVReader.CSVParseException e) {
+        throw new RuntimeException(e); // don't retry if it is parse error
+      } catch (Exception e) { // if it is any other exception, retry may resolve the issue.
+        exception = e;
+        log.info("***Retrying***: {} - {}", fileIdVO, e.getMessage());
+        this.csvReader = openAndSeekCsvReader();
+      }
+    }
+    throw new RuntimeException("***Retried***: Failed, tried " + retryLimit + " times - ", exception);
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (this.header == null) {
+      initHeader();
+    }
+    return this.preLoadedLine != null;
+  }
+
+  @Override
+  public JsonElement next() {
+    if (this.header == null) {
+      initHeader();
+    }
+    JsonElement jsonObject = Utils.csvToJsonObject(this.header, this.preLoadedLine, this.columnSize);
+    this.preLoadedLine = this.nextLine();
+    if (this.preLoadedLine == null) {
+      log.info("----Record count: [{}] for {}", lineCount - 1, fileIdVO);
+    }
+    return jsonObject;
+  }
+
+  private InputStreamCSVReader openAndSeekCsvReader() {
+    String jobId = fileIdVO.getJobId();
+    String batchId = fileIdVO.getBatchId();
+    String resultId = fileIdVO.getResultId();
+    log.info("Fetching [jobId={}, batchId={}, resultId={}]", jobId, batchId, resultId);
+    closeCsvReader();
+    try {
+      InputStream is = conn.getQueryResultStream(jobId, batchId, resultId);
+      BufferedReader br = new BufferedReader(new InputStreamReader(is, ConfigurationKeys.DEFAULT_CHARSET_ENCODING));
+      csvReader = new InputStreamCSVReader(br);
+      for (int j = 0; j < lineCount; j++) {
+        csvReader.nextRecord(); // skip these record
+      }
+      // failed to skip seekLineNumber lines, try to open the file again. same time count+1
+      return csvReader;
```

Comment:
I don't see the sanity check that after skipping the position matches the prior position. This was the old logic.

      for (int i = 0; i < recordsToSkip; i++) {	
        lastCsvRecord = reader.nextRecord();	
      }	

      // make sure the last record processed before the error was the last record skipped so that the next	
      // unprocessed record is processed in the next call to fetchResultBatch()	
      if (recordsToSkip > 0) {	
        if (!this.csvRecord.equals(lastCsvRecord)) {	
          throw new RuntimeException("Repositioning after reconnecting did not point to the expected record");	
        }	
      }	
    }

Reply:
Added it back. test - https://ltx1-holdemaz05.grid.linkedin.com:8443/executor?execid=22307199&job=salesforce_task_full&attempt=0